src/main/java/org/jenkinsci/plugins/parallel_test_executor/ParallelTestExecutor.java

```diff
@@ -143,12 +153,44 @@ public boolean perform(AbstractBuild<?, ?> build, Launcher launcher, BuildListen
         return true;
     }
 
-    static List<InclusionExclusionPattern> findTestSplits(Parallelism parallelism, Run<?,?> build, TaskListener listener, boolean generateInclusions) {
+    static List<InclusionExclusionPattern> findTestSplits(Parallelism parallelism, Run<?,?> build, TaskListener listener,
+                                                          boolean generateInclusions,
+                                                          @CheckForNull final String stageName) {
         TestResult tr = findPreviousTestResult(build, listener);
         if (tr == null) {
             listener.getLogger().println("No record available, so executing everything in one place");
             return Collections.singletonList(new InclusionExclusionPattern(Collections.<String>emptyList(), false));
         } else {
+            Run<?,?> prevRun = tr.getRun();
+            if (prevRun instanceof FlowExecutionOwner.Executable && stageName != null) {
+                FlowExecutionOwner owner = ((FlowExecutionOwner.Executable)prevRun).asFlowExecutionOwner();
+                if (owner != null) {
+                    FlowExecution execution = owner.getOrNull();
+                    if (execution != null) {
+                        DepthFirstScanner scanner = new DepthFirstScanner();
+                        FlowNode stageId = scanner.findFirstMatch(execution, new Predicate<FlowNode>() {
```

Comment:
NIT: I guess creating at least an internal named class like `StageNodeFinderPredicate` or something could have made this slighly less indented and easier to grok :) (+ could possibly make testing this predicate feasible if needed)

Reply:
Sounds reasonable.